Add support to Adaq4224 and ad4030

jonathanns opened the pull request:
## Pull Request Description
This adds AD4030 and ADAQ4224 features (such as PGIA) to ad463x driver and ad463x_fmcz project.
It also adds raw and scale related attributes to IIO.

## PR Type
- [x] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -380,10 +439,53 @@ int32_t ad463x_read_data(struct ad463x_dev *dev,
+int mult_frac (int number, int numerator, int denominator)
+{
+	/** Check if the denominator is not zero to avoid division by zero */
+	double temp;
```
drop please `float/double`.

amiclaus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -161,7 +192,7 @@ int32_t ad463x_spi_reg_write_masked(struct ad463x_dev *dev,
 {
 	if ((mode != AD463X_NORMAL_MODE) && (mode != AD463X_LOW_POWER_MODE))
-		return -1;
+		return -EINVAL;
```
don't do these kind of changes in the same patch. this patch adds a new feature as stated in the commit title/body.

amiclaus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -187,7 +218,7 @@ int32_t ad463x_set_avg_frame_len(struct ad463x_dev *dev, uint8_t mode)
 					   AD463X_AVG_FILTER_RESET);
 	} else {
-		if(mode < 1 || mode > 16)
+		if (mode < 1 || mode > 16)
```
same here. unrelated.

amiclaus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -306,42 +337,66 @@ int32_t ad463x_set_ch_offset(struct ad463x_dev *dev, uint8_t ch_idx,
- * @param [in] dev - ad469x_dev device handler.
- * @param [out] buf - data buffer.
- * @param [in] samples - sample number.
+ * @param dev - ad469x_dev device handler.
```
same here.

amiclaus commented on projects/ad463x_fmcz/src/parameters.h:
```diff
@@ -47,8 +47,8 @@
 /******************************************************************************/
-#define AD463x_DMA_BASEADDR             XPAR_AXI_AD463X_DMA_BASEADDR
-#define AD463x_SPI_ENGINE_BASEADDR      XPAR_SPI_AD463X_SPI_AD463X_AXI_REGMAP_BASEADDR
+#define AD463x_DMA_BASEADDR             XPAR_AXI_AD4X3X_DMA_BASEADDR
```
i think this change breaks the Xilinx build.

jonathanns replied:
how to solve this? since this change is needed to work with the latest HDL build

rbolboac replied:
I will re-run the build, maybe the new HDL artifacts were not yet available in the build system

jonathanns replied:
I talked to the HDL guys and they said to remove this commit and keep as it is on main, so I will do that.

amiclaus reviewed (commented):
Some initial comments that will ease the second round of reivew.
- please clean 6af73c8cb384cd044d29a53fe3c9581ca4671825 from unrelated changes (I left only some comments, but there is more thoughout the commit).
- 7bfcddfd38c2216e0bcabd167894ac04a569e63a lacks commit body.

CiprianRegus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,23 +660,108 @@ int32_t ad463x_init(struct ad463x_dev **device,
+		uint64_t current_diff = abs((int32_t)(target - array[i]));
+
+		if (current_diff < closest_diff) {
+			closest_index = (int32_t)i;
```
you don't need to cast

CiprianRegus commented on drivers/adc/ad463x/ad463x.h:
```diff
@@ -150,7 +150,7 @@
 /**
  * @enum ad463x_id
- * @brief Device type
+ * @brief Device type.
```
unrelated

CiprianRegus commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,23 +660,108 @@ int32_t ad463x_init(struct ad463x_dev **device,
+	ret = no_os_gpio_set_value(dev->gpio_pgia_a0, gain_idx & NO_OS_BIT(0));
+	if (ret != 0)
+		return ret;
+	ret = no_os_gpio_set_value(dev->gpio_pgia_a1, (gain_idx>>1) & NO_OS_BIT(0));
```
replace the bit operations with `no_os_field_get()`

jonathanns replied:
noted.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+	struct iio_ad463x *iio_desc = dev;
+
+	/* Exit register configuration mode */
+	ret = ad463x_exit_reg_cfg_mode(iio_desc->ad463x_desc);
```
`ret` is not checked

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+static int ad463x_iio_read_raw(void *dev, char *buf, uint32_t len,
+			       const struct iio_ch_info *channel, intptr_t priv)
+{
+	uint32_t temp[20];
```
Shouldn't the size be 2 instead?

jonathanns replied:
yes, i increased for tests and forgot to change it back.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+	if(ad463x_dev->has_pgia) {
+		iio_parse_value(buf, IIO_VAL_FRACTIONAL, &vals[0], &vals[1]);
+
+		int32_t gain_idx = ad463x_calc_pgia_gain(vals[0], vals[1], ad463x_dev->vref,
```
declare `gain_idx` at the start of the function.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+
+	int32_t vals[2];
+	/** Scale only configurable if PGIA is available */
+	if(ad463x_dev->has_pgia) {
```
We usually use negative logic in such situations.

```
if (!ad463x_dev->has_pgio)
    return 0;
...
```

jonathanns replied:
Noted.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+
+		int32_t gain_idx = ad463x_calc_pgia_gain(vals[0], vals[1], ad463x_dev->vref,
+				   ad463x_dev->real_bits_precision);
+		return ad463x_set_pgia_gain(ad463x_dev, (uint8_t)gain_idx);
```
You don't need to cast

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+			vals[0] = ad463x_dev->scale_table[i][0];
+			vals[1] = ad463x_dev->scale_table[i][1];
+			iio_format_value(buffer, sizeof(buffer), IIO_VAL_INT_PLUS_NANO, 2, vals);
+			strcat(buf, buffer);
```
`sprintf(buf, "%s ", buffer);`, the you only need 1 function call.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+
+	if(ad463x_dev->has_pgia) {
+		/** Remove data existing in the buffer */
+		strcpy(buf, "");
```
Is there any situation in which we have something in this buffer?

jonathanns replied:
Other drivers do the same, so i just followed. but i will upload a different approach.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+		for (uint8_t i = 0; i < NO_OS_ARRAY_SIZE(ad463x_dev->scale_table); i++) {
+			vals[0] = ad463x_dev->scale_table[i][0];
+			vals[1] = ad463x_dev->scale_table[i][1];
+			iio_format_value(buffer, sizeof(buffer), IIO_VAL_INT_PLUS_NANO, 2, vals);
```
You can use `buf` directly here, thus you don't need to declare `buffer` anymore. Also don't you need to increase the reference to the `buffer` array as you write into it?

jonathanns replied:
not in this case, since Buffer is used to hold a temporary string that will be append to buf and then be overwritten.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.h:
```diff
@@ -62,7 +62,8 @@ struct iio_ad463x {
 };
 
-extern struct iio_device ad463x_iio_desc;
+extern struct iio_device ad463x_iio_desc_one_chn;
```
why do these need to be extern?

jonathanns replied:
Also don't know, just kept as it was before. but i will remove

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
-			if (desc->mask & NO_OS_BIT(ch)) {
-				buff[j++] = data[i];
-			}
+	for (i = 0, j = 0; i < NO_OS_ARRAY_SIZE(data); i++)
+		data[i] = buff[i];
```
`memcpy()`

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
 
-	ret = ad463x_read_data(desc->ad463x_desc, data, nb_samples);
+	/* Exit register configuration mode */
+	ret = ad463x_exit_reg_cfg_mode(desc->ad463x_desc);
```
check `ret`

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,184 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
 	int ret;
-	uint32_t data[nb_samples], i, j, ch;
+	uint16_t total_samples = nb_samples * 2;
+	uint32_t data[total_samples], i, j, ch;
```
It's not usually that great to use the stack for potentially large arrays like that. If you really need the `data` array, allocate the memory on the heap using `no_os_calloc()`.

jonathanns commented:
Updates:
* Unrelated changes removed;
* Dropped double/float usage;
* Unnecessary casts removed;
* Fixed Astyle error and misspellings;
* Improved commit body;
* Used another approach in ad463x_iio_read_scale_avail function;
* Fixed other minor problems stated in the review comments;

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
-		ret = no_os_gpio_direction_output(dev->gpio_resetn, NO_OS_GPIO_LOW);
-		if (ret != 0)
-			return ret;
+	ret = no_os_gpio_direction_output(dev->gpio_resetn, NO_OS_GPIO_LOW);
+	if (ret != 0)
+		goto error_gpio_reset;
```
the previous implementation looks correct to me.
resetn is not allocated in this function so it should not be removed here, simply return without calling the remove function,

jonathanns replied:
Isn't resetn allocated in `no_os_gpio_get_optional` ?  `no_os_gpio_direction_output` and `no_os_gpio_set_value` only fails when allocated, right? 
To summarize: the gpio optional is to keep the driver working even when this GPIOs are not set (in this case nothing will be allocate and it won't return error). but once it is set and allocated, it need to be removed after an error. correct me if I am wrong or if you meant something else.

rbolboac replied:
for some reason I thought dev->gpio_resetn is allocated in ad463x_init (because it is already removed there)...
but I see that it is allocated in  ad463x_init_gpio, so it is alright. I deleted all my comments related to this.

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
-			return ret;
+	no_os_mdelay(100);
+	ret = no_os_gpio_set_value(dev->gpio_resetn, NO_OS_GPIO_HIGH);
+	if (ret != 0)
+		goto error_gpio_reset;
```
same remark as above

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
+	if (dev->has_pgia) {
+		ret = no_os_gpio_get_optional(&dev->gpio_pgia_a0,init_param->gpio_pgia_a0);
+		if (ret != 0)
+			goto error_gpio_reset;
```
return directly

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
+			goto error_gpio_reset;
+		ret = no_os_gpio_get_optional(&dev->gpio_pgia_a1, init_param->gpio_pgia_a1);
+		if (ret != 0)
+			goto error_gpio_reset;
```
in this case gpio_pgia_a0 should be removed

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -380,6 +444,49 @@ int32_t ad463x_read_data(struct ad463x_dev *dev,
+	if (denominator != 0) {
+		int temp = number * numerator;
+		return temp / denominator;
+	} else {
```
else can be removed, simply leave return -EINVAL

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -380,6 +444,49 @@ int32_t ad463x_read_data(struct ad463x_dev *dev,
+ * @param denominator - denominator part of the fraction.
+ * @return Result of the multiplication, or -EINVAL if denominator is zero.
+ */
+int mult_frac(int number, int numerator, int denominator)
```
this could be moved to no_os_util

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -467,17 +576,45 @@ int32_t ad463x_init(struct ad463x_dev **device,
+		dev->has_pgia = true;
+		ad463x_fill_scale_tbl(dev);
+	} else
+		dev->has_pgia = false;
```
brackets are needed in else condition if the if condition has them.

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +665,100 @@ int32_t ad463x_init(struct ad463x_dev **device,
+ * @param size - array size.
+ * @return Gain control index, in case of success, negative number otherwise
+ */
+static int32_t ad463x_find_closest(int32_t target, const int32_t array[],
```
this could be moved to no_os_util

jonathanns replied:
Actualy I found an existing implementation of that. i will check that

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +665,100 @@ int32_t ad463x_init(struct ad463x_dev **device,
+static int32_t ad463x_find_closest(int32_t target, const int32_t array[],
+				   size_t size)
+{
+	if (size == 0) {
```
!size, brackets not needed

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
+
+	/** Configure PGIA pins, if available */
+	if (dev->has_pgia) {
+		ret = no_os_gpio_get_optional(&dev->gpio_pgia_a0,init_param->gpio_pgia_a0);
```
all in all, I would move allocation for gpio_pgia_a0 and gpio_pgia_a1 in init, similar to resetn, and simply handle the gpio value here. it will make the error handling easier.

rbolboac commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,191 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+
+	/* Exit register configuration mode */
+	ret = ad463x_exit_reg_cfg_mode(iio_desc->ad463x_desc);
+	if (ret < 0)
```
I think this driver uses if (ret != 0) syntax, please use this as well

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
+		/* set default values */
+		ret = no_os_gpio_direction_output(dev->gpio_pgia_a0, NO_OS_GPIO_LOW);
+		if (ret != 0)
+			goto error_gpio_pgia;
```
in this case dev->dev->gpio_pgia_a1 is not allocated but you are calling remove for it. this will work only if calloc is used (otherwise you will remove some random memory), and I see this driver uses malloc...
I suggest using calloc and then you can use a single label in which you call all remove functions, if the memory is not allocated, it will be null and the remove function will do nothing.

jonathanns commented:
Updates v2:
* Removed find_closest function and replaced by the no_os_util implementation.
* Removed mult_frac implementation.
* Revert DMA and SPI-Engine base address changes to comply with the newest HDL parameters.
* Fix memory leak handling.
* Other minor fixes.
* Tested again with the latest HDL build.

Observation: Axi_pwm needs to be updated to accommodate the new PWM IP that supports up to 16 channels.

buha commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -53,6 +54,25 @@
+};
+
+/*
+ * Gains stored and computed as fractions to avoid introducing rounding erros.
```
comment typo: `erros`

buha commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,191 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
-			}
+	/** Fill IIO Buffer */
+	for (i = 0, j = 0; i < total_samples; i++)
+		if (desc->mask & NO_OS_BIT(i % 2)) {
```
I would avoid `%` operation in a loop, it can slow execution by a lot. Outside a loop nobody would care, but the ineffectivness is multiplied here by total_samples. If you just wanna check if i is a multiple of 2, only check the least significant bit.

buha replied:
if (i & 0x1)
    odd
else
    even

jonathanns replied:
ok!

buha commented on projects/ad463x_fmcz/Makefile:
```diff
@@ -1,3 +1,8 @@
+IIOD=y
+ADAQ4224=n
```
So i guess no more than one can be `y` at the same time, in which case a variable like AD463X_ID that can take either one of these 3 values would make more sense. This is not critical but I think it's cleaner.

buha reviewed (dismissed):
approving with some optional remarks

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +647,70 @@ int32_t ad463x_init(struct ad463x_dev **device,
+				   gain_idx));
+	if (ret != 0)
+		return ret;
+	ret = no_os_gpio_set_value(dev->gpio_pgia_a1, no_os_field_get(NO_OS_BIT(1),
```
return directly

jonathanns replied:
done.

rbolboac commented on projects/ad463x_fmcz/src/parameters.h:
```diff
@@ -54,13 +54,13 @@
 #define AD463x_SPI_CS                   0
 #define GPIO_OFFSET			54
 #define GPIO_RESETN_1			GPIO_OFFSET + 32
+#define GPIO_PGIA_0			    GPIO_OFFSET + 33
+#define GPIO_PGIA_1			    GPIO_OFFSET + 34
```
fix alignment

rbolboac commented on projects/ad463x_fmcz/src/ad463x_fmc.c:
```diff
@@ -207,12 +265,15 @@ int main()
 		if (ret != 0)
 			return ret;
-		for (i = 0; i < AD463x_EVB_SAMPLE_NO; i+=2)
+		for (i = 0; i < AD463x_EVB_SAMPLE_NO; i+=2) {
```
braces not needed

jonathanns replied:
Done.

rbolboac commented on projects/ad463x_fmcz/src/ad463x_fmc.c:
```diff
@@ -207,12 +265,15 @@ int main()
-		for (i = 1; i < AD463x_EVB_SAMPLE_NO; i+=2)
-			pr_info("ADC sample ch2: %lu : %lu \n", i-1, buf[i]);
+		}
+		for (i = 1; i < AD463x_EVB_SAMPLE_NO; i+=2) {
```
braces not needed

jonathanns replied:
Done.

rbolboac commented on projects/ad463x_fmcz/src/ad463x_fmc.c:
```diff
@@ -207,12 +265,15 @@ int main()
 	/* Read data */
 	while (true) {
 		ret = ad463x_read_data(dev, buf, AD463x_EVB_SAMPLE_NO);
+
```
remove new line

jonathanns replied:
Done.

rbolboac commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,191 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
-	ret = ad463x_read_data(desc->ad463x_desc, data, nb_samples);
-	if(ret)
-		return ret;
+	data = (uint32_t *)no_os_calloc(1, sizeof(uint32_t) * total_samples * 2);
+	if (!data)
+		return -ENOMEM;
```
why are you allocating memory dynamically?
declaring uint32 data[total_samples*2] should work as well

jonathanns replied:
actually it was recommended to do like this in previous reviews.

rbolboac commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,191 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
+	int32_t vals[2];
+	/** Scale only configurable if PGIA is available */
+	if(!ad463x_dev->has_pgia)
+		return 0;
```
not sure returning 0 is the best option, -EINVAL might fit better

jonathanns replied:
Done.

rbolboac reviewed (commented):
some remarks from my side

jonathanns commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -314,26 +350,50 @@ static int32_t ad463x_init_gpio(struct ad463x_dev *dev,
-			return ret;
+	ret = no_os_gpio_direction_output(dev->gpio_resetn, NO_OS_GPIO_LOW);
+	if (ret != 0)
+		return ret;
```
Done.

jonathanns commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +647,70 @@ int32_t ad463x_init(struct ad463x_dev **device,
+ * @param gain_idx - Gain control index, consult values in ad463x_pgia_gain.
+ * @return 0 in case of success, -1 otherwise
+ */
+int32_t ad463x_set_pgia_gain(struct ad463x_dev *dev,
```
Yes, I can do that.

jonathanns commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +647,70 @@ int32_t ad463x_init(struct ad463x_dev **device,
+{
+	int32_t ret;
+	/** Check if gain available in the ADC */
+	if (!dev->has_pgia)
```
done.

jonathanns commented on projects/ad463x_fmcz/src/ad463x_fmc.c:
```diff
@@ -151,44 +171,66 @@ int main()
+	/** Initialize UART when IIO is not enabled */
+	ret = no_os_uart_init(&uart_desc, &iio_uart_ip);
+
+	if (ret <0)
```
done.

jonathanns commented on projects/ad463x_fmcz/src/ad463x_fmc.c:
```diff
@@ -151,44 +171,66 @@ int main()
+#ifndef IIO_SUPPORT
+	/** Initialize UART when IIO is not enabled */
+	ret = no_os_uart_init(&uart_desc, &iio_uart_ip);
+
```
done.

jonathanns commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +647,70 @@ int32_t ad463x_init(struct ad463x_dev **device,
+ * @param precision - Precision  value shifter.
+ * @return Gain control index, in case of success, -1 otherwise
+ */
+int32_t ad463x_calc_pgia_gain(int32_t gain_int, int32_t gain_fract,
```
done.

CiprianRegus commented on drivers/adc/ad463x/iio_ad463x.c:
```diff
@@ -91,31 +151,194 @@ static int32_t _iio_ad463x_prepare_transfer(struct iio_ad463x *desc,
-	ret = ad463x_read_data(desc->ad463x_desc, data, nb_samples);
-	if(ret)
-		return ret;
+	data = (uint32_t *)no_os_calloc(1, sizeof(uint32_t) * total_samples * 2);
```
the first parameter of `calloc()` should be the number of element, and the second one is the size of each element.
So, the proper way to write this would be:

`no_os_calloc(total_samples * 2, sizeof(*data));`

jonathanns replied:
noted, thank you.

rbolboac reviewed (commented):
some nits from my side, but overall looks good... can you confirm when you have tested this successfully?

jonathanns commented:
Updates v3:

* Adressed @rbolboac , @CiprianRegus and @buha comments and sugestions.
* Tested with setup ZedBoard + ADAQ4224.

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -101,6 +121,22 @@ int32_t ad463x_spi_reg_write(struct ad463x_dev *dev,
+ */
+int32_t ad463x_enter_config_mode(struct ad463x_dev *dev)
+{
+	uint8_t buf[3];
```
dev should be checked for null pointer

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -380,6 +444,31 @@ int32_t ad463x_read_data(struct ad463x_dev *dev,
+	unsigned int i;
+	int64_t tmp2;
+
+	val2 = dev->real_bits_precision;
```
dev should be checked for null pointer

rbolboac commented on drivers/adc/ad463x/ad463x.c:
```diff
@@ -529,12 +647,73 @@ int32_t ad463x_init(struct ad463x_dev **device,
+{
+	uint64_t gain_nano, tmp;
+
+	gain_nano = (uint64_t)gain_int * NANO + gain_fract;
```
gain_idx should be checked for null pointer

rbolboac reviewed (approved):
some null checks might be useful... but all in all looks good to me

rbolboac commented:
not sure why build did not start, please perform a force-push (no changes needed) to re-trigger the build

rbolboac commented:
/AzurePipelines run

rbolboac reviewed (approved):
in my opinion, we can merge this.
